resources/queue.py

```diff
@@ -0,0 +1,108 @@
+    exists_find_result = machine_groups.find({"$and": [{"_id": ObjectId(group_id)}, {"queue": {"$exists": True }}]})
+    exists_bool = bool(len(list(exists_find_result)))
+
+    if not exists_bool: # create queue of one person
+        result = machine_groups.update_one({'_id': ObjectId(group_id)},
+                {'$set': {'queue': [str(user_id)]}},
+                upsert=True)
+    else: # add new user to end of array
+        # get array from db
+        queue_result = machine_groups.find({"_id": ObjectId(group_id)},{"queue": 1})
+        queue_array = []
+        for doc in queue_result:
+            queue_array.append(doc)
+        
+        # get queue add append user to the end
+        queue = list(queue_array[0]["queue"])
+        queue.append(str(user_id))
+
+        # update queue in db
+        result = machine_groups.update_one({'_id': ObjectId(group_id)},
+                {'$set': {'queue': queue}},
+                upsert=True)
+
+    return result
+
+# remove user from queue
+def remove_user(machine_groups, group_id, user_id):
+    # get array from db
+    queue_result = machine_groups.find({"_id": ObjectId(group_id)},{"queue": 1})
+    queue_array = []
+    for doc in queue_result:
+        queue_array.append(doc)
+    
+    # get queue remove user from it
+    queue = list(queue_array[0]["queue"])
+    queue.remove(str(user_id))
+
+    # update queue in db
+    if len(queue) == 0:
+        result = machine_groups.update_one({'_id': ObjectId(group_id)},
+                {'$unset': {'queue': ''}},
+                upsert=True)
+    else:
+        result = machine_groups.update_one({'_id': ObjectId(group_id)},
+                {'$set': {'queue': queue}},
+                upsert=True)
+
+    return result
+
+class Queue(Resource):
+    # set the collection to machine_groups
+    def __init__(self, **kwargs):
+        self.db = kwargs['db']
+        self.machine_groups = self.db['machine_groups']
+
+    # general get request to get the queue for a specific machine_group
+    def get(self, search_group):
+        
+        result_cursor = self.machine_groups.find({"_id": ObjectId(search_group)})
+
+        # in order to return a result needs to be {} format
+        return_result = {}
+        
+        for document in result_cursor:
+            return_result[document["name"]] = document["queue"]
+
+        # if unable to find matching query item
+        if return_result == {}:
+            return_result = {'error:' : 'Not Found'}
+
+        return return_result
+        
+    # curl -i -H "Content-Type: application/json" -X POST -d '{"_id":"5c951d991c9d4400008f68e3","user_id":"123259","action":"add"}' http://localhost:5000/queue
+    def post(self):
+        json_data = request.get_json(force=True)
+
+        try:
+            if json_data['action'] == "add":
+                result = add_user(self.machine_groups, json_data['_id'], json_data['user_id'])
+            elif json_data['action'] == "remove":
```

Comment:
Who would be calling remove from the queue? Would this be on the iOS side if they decide to no longer be in the queue? Because if it's for the check in process, the server should probably be the one doing the removal, so we don't need an remove api if that's the case. I am missing some context from the last week tho.

Reply:
The removal could be from various sources theoretically, but the main one I was thinking about was on the HUB side when the user who is next in the queue goes to actually claim the machine. That's why I added an endpoint for it, does that make sense?

Reply:
I don't think the Hubs should be doing this for two reasons.
- The HUBs aren't aware of their machine groups, and I don't think we should force them to ask for the machine group ID.
- The HUBs already send us a user_id and station_id in the machine check in. This is where I think removing from the queue should happen if the user is even allowed up next. It's extra api calls to have the HUBs remove from a queue and check in at the same time when a user taps a machine, when it should all be in one api call.

This call could be good for other systems like the iOS app if they accidentally queue because they are aware of the machine groups IDs. Therefore, I don't think we should get rid of it by any means. I just don't think the HUBs should be using this.